docs/rfc/rfc-026-p2p-bad-peers-checktx.md

```diff
@@ -0,0 +1,375 @@
+  (relevant code [here](https://github.com/cosmos/ibc-go/blob/a0e59b8e7a2e1305b7b168962e20516ca8c98fad/modules/core/ante/ante.go#L23)) 
+- From  Adi: Nodes allow transactions with a wrong `minGas` transaction and gossip them, and other nodes keep rejecting them. (Problem seen with relayers)
+
+### Current state of mempool/p2p interaction
+
+Transactions received from a peer are handled within the [`Receive`](https://github.com/tendermint/tendermint/blob/ff0f98892f24aac11e46aeff2b6d2c0ad816701a/mempool/v0/reactor.go#L158) routine. 
+
+Currently, the mempool triggers a disconnect from a peer in the case of the following errors:
+
+  - [Unknown message type](https://github.com/tendermint/tendermint/blob/ff0f98892f24aac11e46aeff2b6d2c0ad816701a/mempool/v0/reactor.go#L184)
+
+However, disconnecting from a peer is not the same as banning the peer. The p2p layer will close the connecton but 
+the peer can reconnect without any penalty, and if it as a persistent peer, a reconnect will be initiated
+from the node. 
+
+### Current support for peer banning
+
+The p2p layer implements only banning peers for an indefinite amount of time, by marking them
+as bad (calling the `MarkBad` routine implemented by the `Switch`). The peers are reinstated to the list 
+of good peers only if the node requires more peers. They can also not be marked as bad forever (blacklisted). 
+
+The application can blacklist peers via ABCI if the 
+[`filterPeers`](../../spec/abci/abci%2B%2B_app_requirements.md#peer-filtering) 
+config flag is set, by providing a set of peers to ban to Tendermint. 
+
+If the discussion in this RFC deems a different banning mechanism is needed,
+the actual implementation and design of this mechanism will be discussed in a separate RFC.
+This mechanism should be generic, designed within the p2p layer and simply provide an interface
+for reactors to indicate peers to ban and for how long. It should not involve any mempool
+specific design considerations. 
+
+## Discussion
+
+If this feature is to be implemented we need to clearly define the following:
+1. What does banning a peer mean:
+   1. A peer can be simply disconnected from. 
+   2. Peer is disconnected from and banned.
+   3. Conditions for the peer to be banned.
+2. If `CheckTx` signals a peer should be banned, how do we know which peer(s) to ban?
+3. Are there possible attack scenarios by allowing this. 
+
+Any further mentions of `banning` will be agnostic to the actual way banning is implemented by the p2p layer. 
+
+
+### 1. What does banning a peer mean 
+
+Tendermint recognizes that peers can accept transactions into their mempool as valid but then when the state changes, they can become invalid. There are also transactions that are received that could never have been valid (for examle due to misconfiguration on one node). 
+We thus differentiate two scenarios - a) where `CheckTx` fails due to reasons already 
+known and b) where `CheckTx` deems a transactions could never have been valid.
```

Comment:
In general, `CheckTx` should be stateful , in the sense that it compares the current transaction against the latest state. The application is the one that defines whether the failure is something that happened due to a difference in states (height etc.) or is the failure a reason for banning. THese two distinctions cover the case where the first scenario happens very often, while the other is immediate reason for banning.

Reply:
> In general, CheckTx should be stateful

Not necessarily. The SDK only has one stateful check in the antehandler (that I know of) which is to check the nonce numbers everything else about it is stateless (It is essentially one large `ValidateBasic` function which mostly ensures that the bytes match a recognised protobuf message).

It doesn't make sense to spend too much time checking against state because you don't know whether that's the state that the transactions will eventually execute against so you will need to do all those stateful checks again. 

CheckTx should be seen as an initial gatekeeper for transactions but not the final gatekeeper. The final gatekeeper is `PrepareProposal` and `ProcessProposal`.

Reply:
This is the source for my comment btw: https://github.com/tendermint/tendermint/blob/main/spec/abci/abci%2B%2B_methods.md#checktx . We might need to rephrase this if you feel that's needed.